dask_expr/merge.py

```diff
@@ -0,0 +1,173 @@
+        #  2. Add multi-partition broadcast merge
+        #  3. Add/leverage partition statistics
+
+        # Check for "trivial" broadcast (single partition)
+        npartitions = max(left.npartitions, right.npartitions)
+        if (
+            npartitions == 1
+            or left.npartitions == 1
+            and how in ("right", "inner")
+            or right.npartitions == 1
+            and how in ("left", "inner")
+        ):
+            return BlockwiseMerge(left, right, **self.kwargs)
+
+        # Check if we are merging on indices with known divisions
+        merge_indexed_left = (
+            left_index or _contains_index_name(left, left_on)
+        ) and left.known_divisions
+        merge_indexed_right = (
+            right_index or _contains_index_name(right, right_on)
+        ) and right.known_divisions
+
+        # NOTE: Merging on an index is fragile. Pandas behavior
+        # depends on the actual data, and so we cannot use `meta`
+        # to accurately predict the output columns. Once general
+        # partition statistics are available, it may make sense
+        # to drop support for left_index and right_index.
+
+        shuffle_left_on = left_on
+        shuffle_right_on = right_on
+        if merge_indexed_left and merge_indexed_right:
+            # fully-indexed merge
+            if left.npartitions >= right.npartitions:
+                right = Repartition(right, new_divisions=left.divisions, force=True)
+            else:
+                left = Repartition(left, new_divisions=right.divisions, force=True)
+            shuffle_left_on = shuffle_right_on = None
+
+        # TODO:
+        #   - Need 'rearrange_by_divisions' equivalent
+        #     to avoid shuffle when we are merging on known
+        #     divisions on one side only.
+        #   - Need mechanism to shuffle by an un-named index.
+        else:
+            if left_index:
+                shuffle_left_on = left.index._meta.name
+                if shuffle_left_on is None:
+                    raise NotImplementedError("Cannot shuffle unnamed index")
+            if right_index:
+                shuffle_right_on = right.index._meta.name
+                if shuffle_right_on is None:
+                    raise NotImplementedError("Cannot shuffle unnamed index")
+
+        if shuffle_left_on:
+            # Shuffle left
+            left = Shuffle(
+                left,
+                shuffle_left_on,
+                npartitions_out=npartitions,
+                backend=shuffle_backend,
+            )
+
+        if shuffle_right_on:
+            # Shuffle right
+            right = Shuffle(
+                right,
+                shuffle_right_on,
+                npartitions_out=npartitions,
+                backend=shuffle_backend,
+            )
+
+        # Blockwise merge
+        return BlockwiseMerge(left, right, **self.kwargs)
+
+
+class BlockwiseMerge(Merge, Blockwise):
+    """Blockwise merge class"""
```

Comment:
This docstring seems noninformative.  

Totally ok for now while we're early, just calling it out as something that'll probably have to change before we think about bringing more folks on.  Probably as we get closer to thinking of this as a real project we'll want things to start looking more polished.

We're not there yet, but I think it's probably useful to start thinking about this.